【Task27】Addition of local storage references on page transitions

csakyo opened the pull request:
# [Task27]
https://github.com/kenmori/handsonFrontend/blob/master/work/markup/1.md#27

## [codesandbox](https://codesandbox.io/s/shy-architecture-74ci4k) / Latest commit: https://github.com/csakyo/homework/commit/f52303a0d4a60f54d41ea8426fa3cfd9709ceb40

## 課題27仕様
index.htmlページ遷移時に 学習の為のローカルストレージのトークンを参照して もし {token: far0fja*ff]afaawfqrlzkfq@aq9283af}の値があれば 作ったコンテンツUI画面を ローカルストレージにそれがなければログイン画面(login.html)に遷移するように作ってください

ローカルストレージの有無はコンテンツページ遷移時に確認してください
・開発ツールのApplication -> localStorageの中の値を削除してリロードした際もログイン画面に遷移するはずです
・またログイン画面からurlをコンテンツページのそれに変更してもちゃんと挙動ができていること

◆The following user data were set as fixed values this time.
user name : tanaka
user email : tanaka@gmail.com
pass : N302aoe3

## [Comment for this Task]
I added local storage token reference in index.html(NewsUI page). 
The only differences from Issue 26 in this issue are the parts I have commented on. If I have missed something about the operation or specifications, we would appreciate it if you could point it out to us.

csakyo committed:
Feat: Add local storage reference when transitioning to content page

csakyo commented on question27/js/news.js:
```diff
@@ -0,0 +1,158 @@
+
+if(!localStorage.getItem('token')) {
+  window.location.href = "./login.html"
+}
```
I added check for tokens in local storage on transition to index.html.

kenmori commented on question27/js/login.js:
```diff
@@ -0,0 +1,114 @@
+'use strict';
+
+import { Chance } from "chance";
+const chance = new Chance();
```
👍

nakagawamai replied:
I've never heard of Chance.js before!
I see that it can generate random values.
Thanks a lot for the learning.

kenmori commented on question27/js/login.js:
```diff
@@ -0,0 +1,114 @@
+  window.location.href = "./index.html";
+}
+
+const isValidStatus = {
+  name_mail: false,
+  password: false,
+}
```
ローカルステートですね 👍

haru-programming commented on question27/js/news.js:
```diff
@@ -0,0 +1,158 @@
+  }
+  finally {
+    console.log('CONPLETE!');
+  }
```
Q.
Is this ’finally’ necessary?
I would appreciate it if you could tell me!

csakyo replied:
Thank you for your review!! I am sorry you told me before :bow: Fixed so please confirm.

fix commit : https://github.com/csakyo/homework/pull/41/commits/98d73171be17edd87be5b240a56cba6968730b46

haru-programming commented on question27/css/style.css:
```diff
@@ -0,0 +1,222 @@
+.submit_button:disabled {
+    background: #ccc;
+    cursor: not-allowed;
+}
```
IMO
I will share what I have learned so far from the review.
Normally, after you type everything in the input, you have to click somewhere else for the validation to go through, but with `pointer-events`, you can press the submit button directly after you type :)

```suggestion
.submit_button {
    width: 100%;
    background: #000;
    font-size: 1rem;
    line-height: 1.2;
    font-weight: 300;
    height: 60px;
    color: #fff;
    border-radius: 30px;
    cursor: pointer;
    pointer-events: auto; //here
}
.submit_button:disabled {
    background: #ccc;
    cursor: not-allowed;
    pointer-events: none; //here
}
```

csakyo replied:
Thank you for your review! This is very convenient! Thanks for letting me know! 😄 

fixed commit : https://github.com/csakyo/homework/pull/41/commits/456f01c744fd68b9bd4eb394b0e403c4b90d0c84

nakagawamai commented:
nits
The behavior of the logout button below was a bit unsettling from the user's point of view, so it might be a good idea to add a process to hide it on screen transitions.
・When screen transitioning to index.html, the logout button appears before the content.
・When logging out from  index.html and transitioning to login.html, the logout button remains displayed for a while.

https://user-images.githubusercontent.com/99260576/197397771-1c3ae20e-67be-44a8-89a3-47ce9d45d62f.mov

https://user-images.githubusercontent.com/99260576/197397719-b2ac3fe7-7751-4b7f-bbb7-78e14a5731f8.mov

sae-github commented on question27/js/login.js:
```diff
@@ -0,0 +1,114 @@
+  try {
+    const serverResponseData = await userDataVerification();
+    const token = serverResponseData.token; 
+    localStorage.setItem("token", JSON.stringify(token));
```
IMO
I think it is better to run the token set in a different location.
Because I think that doing the login verification and setting the results to localStorage are two different jobs.

csakyo replied:
Thank you for your review!
You are right. How about moving the token set to init()?  Please confirm.

fixed commit : 
https://github.com/csakyo/homework/pull/41/commits/cc1f84d857682e84ce9f38302b2817d20c3af2ff
https://github.com/csakyo/homework/pull/41/commits/42f9e38fe5ee08d2bec854c7a90591acc2c9c766

csakyo committed:
Refactor: The logout button was changed to be inserted from JS.

csakyo committed:
Replacing images

csakyo committed:
Refactor: Remove unnecessary FINALLY statements

csakyo committed:
Refactor: Change pointer event

csakyo commented:
> nits The behavior of the logout button below was a bit unsettling from the user's point of view, so it might be a good idea to add a process to hide it on screen transitions. ・When screen transitioning to index.html, the logout button appears before the content. ・When logging out from index.html and transitioning to login.html, the logout button remains displayed for a while.
> 
>  2022-10-23.23.22.03.mov 
>  2022-10-23.23.21.12.mov

@nakagawamai 
Thank you for your review! You are right, I noticed that when you said that.
To match the movement with the others, I decided to insert a logout button from the JS side. I have modified it and hope you will check it.

fixed commit : https://github.com/csakyo/homework/pull/41/commits/b64bd2e9ac686580a57c0a030f8607daed405709

csakyo committed:
Refactor: Move local storage set to init function

csakyo committed:
Refactor: Move local storage set to init function

sae-github commented on question27/js/login.js:
```diff
@@ -0,0 +1,113 @@
+const init = async () => {
+  const verificationResult = await loginVerification();
+  const token = verificationResult.token;
+  localStorage.setItem("token", JSON.stringify(token));
```
MUST
The token has been stored in localStorage regardless of the result of user authentication.
Therefore, if you visit index.html after a failed login attempt, you will be redirected to the content page.

csakyo replied:
Thank you for your review! I have completely overlooked this... I added if statement in init(). Please confirm the correction.

fixed commit : https://github.com/csakyo/homework/pull/41/commits/d741f6c55c54fd89fc5f23ee47e2aae128ffd2c1

csakyo committed:
Refactor: Added if statement to init function because it is required as a condition for login

csakyo committed:
Refactor: Move the script for checking tokens to HTML.

nakagawamai commented on question27/js/news.js:
```diff
@@ -47,6 +48,23 @@ init();
+  logoutButtonArea.appendChild(logoutButton);
+}
+
+const logoutButtonEvent = () => {
```
IMO
I think the function name should contain a verb.

csakyo replied:
Thank you for your review! You are right! I fixed the function name but what do you think? Please confirm.

fixed commit : https://github.com/csakyo/homework/pull/41/commits/6464546d7371a1b73aac51c0af15d8702bb1c525

nakagawamai replied:
Thank you for your correction.

I think that the meaning is conveyed, so I think it's good,
I thought something like `setEventForLogoutButton` or `addEventForLogoutButton` would be more appropriate.
It may be a matter of taste, so I will approved.

csakyo committed:
Refactor:Change function name to something appropriate

csakyo committed:
Refactor:Change function name to something appropriate
